src/mongocxx/database.cpp

```diff
@@ -157,5 +164,16 @@ collection database::collection(stdx::string_view name) const {
     return mongocxx::collection(*this, name);
 }
 
+const database::impl& database::get_impl() const {
+    if(!_impl) {
+        throw operation_exception{make_error_code(error_code::k_database_uninitialized)};
+    }
+    return *_impl;
+}
+
+database::impl* database::get_impl() {
+    return &(const_cast<database::impl&>(static_cast<const database&>(*this).get_impl()));
```

Comment:
Quite an expression!

I think it could work better as multiple lines. I'm assuming here you change the return type to & to match the const get_impl

```
auto cthis = const_cast<const database*>(this);
return const_cast<database::impl&>(cthis->get_impl());
```